feat(chart): prevent CRDs from being recycled

cmontemuino opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->


## Description

With current implementation a chart uninstall also removes the CRDs, which triggers deletion of all created CRs.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Add a configuration to keep CRDs by default, instructing Helm to not delete such resources. Also add the possibility to specify extra annotations. This is especially interesting for Argo CD users, as Argo CD does not install charts the Helm way. In such a case, an extra annotation to prevent prunning resources is required.

## Regression
<!-- Is this PR fixing a regression? (Yes / No) -->

<!-- If Yes, optionally please include version or commit id or PR# that caused this regression, if you have these details. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

**Case 1: keep all CRDs**

```shell
helm chart template
```

**Case 2: keep CSI VolumeSnaphots CRDs only**

```shell
helm chart template --set crds.jaeger.keep=false
```

**Case 3: keep Jaeger CRDs only**

```shell
helm chart template --set crds.csi.volumeSnapshots.keep=false
```

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.

cmontemuino committed:
feat(chart): prevent CRDs from being recycled

Problem: with current implementation a chart uninstal also removes the
CRDs, which triggers deletion of all created CRs.

Solution: add a configuration to keep CRDs by default, instructing Helm
to not delete such resources. Also add the possibility to specify extra
annotations. This is especially interesting for Argo CD users, as Argo
CD does not install charts the Helm way. In such a case, an extra
annotation to prevent prunning resources is required.

Signed-off-by: cmontemuino <1761056+cmontemuino@users.noreply.github.com>

tiagolobocastro commented on chart/charts/crds/templates/csi-volume-snapshot.yaml:
```diff
@@ -7,6 +7,7 @@ metadata:
   annotations:
```
Should we also remove the crdName check above? That was added post your change because upgrades would otherwise fail:
> Error: UPGRADE FAILED: Unable to continue with update: CustomResourceDefinition "volumesnapshotclasses.snapshot.storage.k8s.io" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; label validation error: missing key "app.kubernetes.io/managed-by": must be set to "Helm"; annotation validation error: missing key "meta.helm.sh/release-name": must be set to "mayastor"; annotation validation error: missing key "meta.helm.sh/release-namespace": must be set to "mayastor"

I think it was maybe not a good approach though, as it's perhaps better to simply document that for upgrades a user should disable these with --set crds.enabled=false?

cmontemuino replied:
I also believe it's better to remove the `crdIsAbsent` check everywhere. I'll proceed with the change.

cmontemuino replied:
addressed in 4ab5529064e4623dcbf5510a0795043c25cc5979

cmontemuino committed:
fix(chart): remove crdIsAbsent check

Problem: the crdIsAbsent check is preventing to upgrade CRDs

Solution: remove the check and rely on `crds.xxx.enabled` property
instead. This should be a conscious decicion made by users anyways.

Signed-off-by: cmontemuino <1761056+cmontemuino@users.noreply.github.com>

tiagolobocastro commented:
thanks @cmontemuino!
bors merge

tiagolobocastro commented:
@cmontemuino could you please re-run the helm doc generator, here's the failure from CI:
```
 ## Values

 

-| Key | Type | Default | Description |

-|-----|------|---------|-------------|

-| csi.volumeSnapshots.annotations | object | `{}` | Annotations to be added to all CRDs |

-| csi.volumeSnapshots.enabled | bool | `true` | Install Volume Snapshot CRDs |

-| csi.volumeSnapshots.keep | bool | `true` | Keep CRDs on chart uninstall |

-| jaeger.annotations | object | `{}` | Annotations to be added to all CRDs |

-| jaeger.enabled | bool | `true` | Install Jaeger CRDs |

-| jaeger.keep | bool | `true` | Keep CRDs on chart uninstall |

+| Key | Description | Default |

+|:----|:------------|:--------|

+| csi.&ZeroWidthSpace;volumeSnapshots.&ZeroWidthSpace;annotations | Annotations to be added to all CRDs | <pre>{<br><br>}</pre> |

+| csi.&ZeroWidthSpace;volumeSnapshots.&ZeroWidthSpace;enabled | Install Volume Snapshot CRDs | `true` |

+| csi.&ZeroWidthSpace;volumeSnapshots.&ZeroWidthSpace;keep | Keep CRDs on chart uninstall | `true` |

+| jaeger.&ZeroWidthSpace;annotations | Annotations to be added to all CRDs | <pre>{<br><br>}</pre> |

+| jaeger.&ZeroWidthSpace;enabled | Install Jaeger CRDs | `true` |

+| jaeger.&ZeroWidthSpace;keep | Keep CRDs on chart uninstall | `true` |
```

cmontemuino committed:
docs: generate documentation

Signed-off-by: cmontemuino <1761056+cmontemuino@users.noreply.github.com>

Abhinandan-Purkait commented:
bors try

Abhinandan-Purkait commented:
Thanks @cmontemuino. Are we good to merge now?

Abhinandan-Purkait commented:
bors merge
